pm/sigverifier.go

```diff
@@ -0,0 +1,58 @@
+package pm
+
+import (
+	"fmt"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// SigVerifier is an interface which describes an object capable
+// of verification of ECDSA signatures produced by ETH addresses
+type SigVerifier interface {
+	// Verify checks if a provided signature over a message
+	// is valid for a given ETH address
+	Verify(addr ethcommon.Address, sig []byte, msg []byte) bool
+}
+
+// ApprovedSigVerifier is an implementation of the SigVerifier interface
+// that relies on an implementation of the Broker interface to provide a registry
+// mapping ETH addresses to approved signer sets. This implementation will
+// recover a ETH address from a signature and check if the recovered address
+// is approved
+type ApprovedSigVerifier struct {
+	broker Broker
+}
+
+// NewApprovedSigVerifier returns an instance of an approved signature verifier
+func NewApprovedSigVerifier(broker Broker) *ApprovedSigVerifier {
+	return &ApprovedSigVerifier{
+		broker: broker,
+	}
+}
+
+// Verify checks if a provided signature over a message
+// is valid for a given ETH address
+func (sv *ApprovedSigVerifier) Verify(addr ethcommon.Address, sig []byte, msg []byte) (bool, error) {
+	personalMsg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", 32, msg)
+	personalHash := crypto.Keccak256([]byte(personalMsg))
+
+	pubkey, err := crypto.SigToPub(personalHash, sig)
+	if err != nil {
+		return false, err
+	}
+
+	rec := crypto.PubkeyToAddress(*pubkey)
+
+	if addr == rec {
+		// If recovered address matches, return early
+		return true, nil
+	}
+
+	approved, err := sv.broker.IsApprovedSigner(addr, rec)
```

Comment:
I'm wondering about the performance implications this will have. Under the following conditions I think this might be problematic:

- If we assume a naive implementation, this would result in a function call to an ETH node every time, which could take a while
- If ticket validation is a blocker for processing a segment

Will these conditions be true when we implement `Broker` and integrate ticket validation in the `ServeSegment` flow?

Reply:
Great point. Here is a summary of an offline discussion about this:

An ETH call to check for approved signers to validate every ticket would create a performance hit for ticket validation which affects segment processing performance. In the short term for MVS, we can remove the approved signer feature client side and instead the recipient will just check that the signature corresponds to the sender address in the ticket (only requires client side cryptographic operations). In the longer term, we could implement either an caching layer that replicates state from the on-chain contract so that a recipient can access a sender's approved signers without an ETH call.

Reply:
>  In the longer term, we could implement either an caching layer

One way is to to lookup the list when the first segment of a stream comes in. That list only needs to be retained for the duration of the stream (eg, transcode loop). There'd be a hit for the very first segment, but subsequent lookups can remain local.

I wonder if borrowing the Explorer's GraphQL caching layer would also help with the initial lookup.

Reply:
GraphQL is definitely worth exploring for this use case. We just need to make sure it has sufficiently efficient response times and no hangups, since we don't want GraphQL to be the source of segment processing delays :)